File: project_euler/Problem089/Problem089.java
package project_euler;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;

public class Problem089 {
    public void solution() throws IOException {
        String fileName = "roman.txt";
        int result = 0;
        try (BufferedReader handle = new BufferedReader(new InputStreamReader(new FileInputStream(fileName)))) {
            String line;
            while (null != (line = handle.readLine())) {
                char previous = ' ';
                char pivot = ' ';
                int count = 1;
                for (char value : line.toCharArray()) {
                    if (pivot == value) {
                        ++count;
                    } else {
                        previous = pivot;
                        pivot = value;
                        count = 1;
                    }

                    if (count >= 4) {
                        if (('I' == pivot && 'V' == previous) || ('X' == pivot && 'L' == previous) || ('C' == pivot && 'D' == previous)) {
                            result += 3;
                        } else if ('M' != pivot) {
                            result += 2;
                        }
                    }
                }
            }
        }
        System.out.println(result);
    }
}

File: project_euler/Problem089/Problem089.py
class Problem089:
    def solution(self) -> None:
        file_name, result = 'roman.txt', 0
        with open(file_name, 'r') as handle:
            for line in handle:
                previous, pivot, count = ' ', ' ', 1
                for value in line:
                    if pivot == value:
                        count += 1
                    else:
                        previous, pivot, count = pivot, value, 1

                    if count >= 4:
                        if ('I' == pivot and 'V' == previous) or ('X' == pivot and 'L' == previous) or ('C' == pivot and 'D' == previous):
                            result += 3
                        elif 'M' != pivot:
                            result += 2
        print(result)
